### DISCUSSION TITLE
Is there a possibility to add an additional field to all value fields

### DISCUSSION THREAD

**Discussion by eduardRedberry** [react-querybuilder/react-querybuilder#296] [Q&A]
Hello,

Is there a possibility to add an additional field to all value fields, like to have two value fields?

Thank you in advance,

**Comment by jakeboone02** [comment]
Hi @eduardRedberry, yes that is possible. You'll need to define a custom `valueEditor` component.

TL;DR: [This CodeSandbox example](https://codesandbox.io/s/ooh2st) demonstrates the technique. The relevant code is also copied below.

Longer answer:

[This page](https://react-querybuilder.js.org/docs/tips/custom-with-fallback) in the Tips & Tricks section of the documentation site should help you understand how to set up a custom value editor that also has a fallback to the default `ValueEditor` for rules that don't need custom functionality. (One of the examples on that page uses a date range selector, which uses a single input but handles two values.)

For your use case, you'll need to keep in mind that each rule only has one standard `value` property. That means that the only supported ways to store multiple values are to make the `value` property either an object (e.g. `{ value: { value1: 'hello', value2: 'world' } }`) or an array (e.g. `{ value: ['hello', 'world'] }`).

There are two gotchas to keep in mind when using an object or array as the `value`:

1. The default `value` for new rules is an empty string (`''`). In order to avoid having to handle the case where a rule's `value` is not already an array, you should set the `defaultValue` property on all relevant [fields](https://react-querybuilder.js.org/docs/typescript#fields) (or use the [`getDefaultValue` prop](https://react-querybuilder.js.org/docs/api/querybuilder#getdefaultvalue) on the `<QueryBuilder />` component).
2. When one of the inputs fires its `onChange` handler, you only want to update the array element or object property associated to that input, leaving the rest alone. For example, when the user updates text in the first text box, you want the first element of the array to update, but the second element should stay the same. Vice versa for the second text box.

```tsx
import { useState } from 'react';
import {
  Field,
  QueryBuilder,
  RuleGroupType,
  ValueEditor,
  ValueEditorProps,
} from 'react-querybuilder';
import 'react-querybuilder/dist/query-builder.scss';

const fields: Field[] = [
  { name: 'firstName', label: 'First Name' },
  { name: 'lastName', label: 'Last Name' },
  // Notice the defaultValue declaration on the next line:
  { name: 'fieldWithTwoValues', label: 'Two Values', defaultValue: ['', ''] },
];

const CustomValueEditor = (props: ValueEditorProps) => {
  if (props.field === 'fieldWithTwoValues') {
    return (
      <span className={props.className} style={{ display: 'flex' }}>
        <input
          style={{ width: 80 }}
          value={props.value[0]}
          onChange={(e) =>
            // Here we update the first element and just echo the second
            props.handleOnChange([e.target.value, props.value[1]])
          }
        />
        <input
          style={{ width: 80 }}
          value={props.value[1]}
          onChange={(e) =>
            // Here we do the reverse: update the second element and echo the first
            props.handleOnChange([props.value[0], e.target.value])
          }
        />
      </span>
    );
  }

  // Only rules for the field "fieldWithTwoValues" will have two inputs.
  // The rest will fall back to the default ValueEditor.
  return <ValueEditor {...props} />;
};

export default function App() {
  const [query, setQuery] = useState<RuleGroupType>({
    combinator: 'and',
    rules: [
      {
        field: 'fieldWithTwoValues',
        operator: '=',
        value: ['Value 1', 'Value 2'],
      },
      {
        field: 'firstName',
        operator: '=',
        value: '',
      },
      {
        field: 'lastName',
        operator: '=',
        value: '',
      },
    ],
  });

  return (
    <div>
      <QueryBuilder
        fields={fields}
        query={query}
        onQueryChange={(q) => setQuery(q)}
        controlElements={{
          valueEditor: CustomValueEditor,
        }}
      />
    </div>
  );
}
```

**Reply by jakeboone02** [reply]
Hi @eduardRedberry. If the response above helped you out, could you mark it as the answer?

**Comment by eduardRedberry** [comment]
Hi @jakeboone02, thank you a lot for your answer.

Could you please say whether I can add like that additional field?
![image](https://user-images.githubusercontent.com/100134431/165249992-fc9c88a7-963f-4acf-977c-6f276e7aa7ac.png)

**Reply by eduardRedberry** [reply]
like to have `{ field: '', aggregation: ''  operator: '',   value: '' }`
if the above structure isn't possible at this point to have at least like `{ field: ['',''],  operator: '',   value: '' }`

**Reply by jakeboone02** [reply]
There are a couple of ways to do this.

1. Add an `aggregation` property to rule objects (as you suggested). By default, a `ValueEditor` component can only update the `value` property of its rule (via the `handleOnChange` callback), so this method would require you to use the `update` helper function to handle the entire query object instead. It also requires you to manage state external to the component, since you cannot define the custom component inside the main component.
2. Store an object as the `value` property that itself includes `aggregation` and `value` properties.

[This CodeSandbox](https://codesandbox.io/s/59x264) demonstrates both methods.
- Select the **"Agg. Rule Property"** field to see the first method in action. (I used Redux to manage state.)
- Select the **"Agg. Value Property"** field to see the second method in action.

The difference can be seen in the sample query object below. The first rule has the `aggregation` property as part of the rule itself, and the second rule has the `aggregation` property in the object assigned to the `value` property.

```json
{
  "combinator": "and",
  "rules": [
    {
      "field": "aggRuleProperty",
      "operator": "=",
      "value": "Value 1, Value 2",
      "aggregation": "sum"
    },
    {
      "field": "aggValueProperty",
      "operator": "=",
      "value": {
        "aggregation": "concat",
        "value": "1, 2, 3"
      },
      "aggregation": "sum"
    }
  ],
  "not": false
}
```

Hope that helps!

**Reply by jakeboone02** [reply]
It just occurred to me that you could store both the aggregation type and the value in the same string and then parse them out. This allows you to use the standard `handleOnChange` callback in a more straightforward way.

I've updated the [CodeSandbox](https://codesandbox.io/s/59x264) with an "Agg. Parse Property" field that uses this technique.
